Feat/bootcamp paging 부트캠프 전체 조회 및 검색 기능 추가

shCHO9801 opened the pull request:
### 변경사항
closes : #44 
**AS-IS**
- 부트캠프 목록을 단순 전체 조회하는 기능만 존재
- 필터, 키워드 검색, 정렬 기능이 없음
- 정적 컨트롤러 구현만 존재하며 동적 쿼리 적용 어려움

**TO-BE**
- 부트캠프 전체 조회 API /api/bootcamps
    - 정적 필터 조건 추가
        - 지역 (region)
        - 카테고리 (category)
        - 평균 평점 (minRating) - 평점대 기준 (예: 3 → 3.0 ~ 3.9)
        - 기간 (duration) - 4주 미만, 4~12주, 12주 이상
    - 페이징 처리 적용 (PagedResponseDto)
- 부트캠프 검색 API /api/bootcamps/search
    - 키워드 검색 기능 추가 (keyword)
        - 부트캠프 이름 기준 검색
    - 정렬 기능 추가 (sort)
        - 최신순 (latest)
        - 평점순 (rating)
        - 리뷰순 (popular)
- QueryDSL 기반으로 동적 조건 처리
- 전체 구조를 컨트롤러 / 서비스 / 쿼리 리포지토리로 분리하여 확장성 확보

### 테스트
- [x]  API 테스트 (Postman으로 전체 조회 및 검색 기능 정상 응답 확인)
- [x] 필터 조합 (카테고리 + 지역 + 평점 + 기간) 동작 확인
- [x] 정렬 조건 (latest / rating / popular) 응답 확인
- [x] 키워드 검색시 부분 일치 검색 확인 (대소문자 구분 없음)
![스크린샷 2025-04-05 오전 4 40 15](https://github.com/user-attachments/assets/545cee2d-5877-49cb-a516-02356ae3ab3e)
![스크린샷 2025-04-05 오전 4 40 30](https://github.com/user-attachments/assets/d65f547a-d597-4850-afb7-a8a7d65244df)

shCHO9801 committed:
Merge branch 'develop' into feat/bootcamp-paging

shCHO9801 committed:
Merge remote-tracking branch 'origin/develop' into feat/bootcamp-paging

shCHO9801 committed:
feat : 부트캠프 검색, 필터 조회 추가

geonuu09 commented on boottalk/src/main/java/com/icandoit/boottalk/bootcamp/controller/BootcampController.java:
```diff
@@ -29,19 +28,49 @@ public class BootcampController {
+	// 검색 기반 조회 (키워드 + 필터 + 정렬 포함)
+	@GetMapping("/search")
+	public ResponseEntity<PagedResponseDto<BootcampResponseDto>> searchBootcamps(
+		@RequestParam(required = false) String region,
```
RequestParam(required = false)로 되어 있는데, 값들이 입력되지 않더라도 default 값은 있어야 하지 않을까 싶습니다! 
sort를 string으로 받는 이유가 있을지 궁금합니다! 
만약 param에 "LATEST"가 들어가면 오류가 생기진 않나요?

shCHO9801 replied:
@geonuu09 
region, category, minRating, duration, keyword는 필터링 조건으로 사용자가 선택적으로 전달하는 값입니다.
따라서 전달되지 않은 경우에는 필터링 없이 전체 검색이 되도록 설계되어 있어 null 값을 허용합니다.
예를 들어 region이 없으면 지역 필터 없이 모든 부트캠프가 조회됩니다.

정렬은 현재 rating, popular, latest 세 가지 문자열 값으로 받고 있으며, QueryDsl에서 switch로 처리 중입니다.

다만 sort 파라미터에 "LATEST"와 같이 대문자가 들어오면 현재 switch에서 "latest"로 비교하기 때문에 매칭되지 않아 기본값인 latest로 fallback 됩니다.

이 문제를 방지하기 위해 sort.toLowerCase()로 소문자 처리를 추가하는 것으로 해결이 가능할거 같은데 어떠신가요??

geonuu09 replied:
아하 그렇군요! 
문제가 생기지 않는다면 성현님이 진행하고자 하는 방향으로 진행하셔도 좋을 것 같아요!

geonuu09 reviewed (approved):
고생하셨습니다 👍🏻

shCHO9801 committed:
Merge remote-tracking branch 'origin/develop' into feat/bootcamp-paging

shCHO9801 committed:
refactor : sort 값을 소문자로 변환하여 대소문자 구분없이 사용 가능하게 수정

shCHO9801 committed:
test: BootcampService 테스트 수정 및 추가

- BootcampServiceTest
  - 부트캠프 ID 조회 테스트에서 Course 정보 누락으로 인한 NPE 해결 (Course mock 추가)
  - 검색 기능 테스트 추가 (필터/키워드 포함된 경우, 빈 결과 반환 테스트 포함)

P-Taeyoung reviewed (commented):
고생하셨습니다~
